docs/topics/tasks.txt

```diff
@@ -0,0 +1,337 @@
+These arguments are provided as additional keys in the :setting:`TASKS` setting.
+Valid arguments are as follows:
+
+* :setting:`ENQUEUE_ON_COMMIT <TASKS-ENQUEUE_ON_COMMIT>`: Whether a task should
+  be enqueued at the end of the current transaction (if there
+  is one) commits successfully, rather than enqueueing immediately. This argument
+  defaults to ``True``.
+
+* :setting:`QUEUES <TASKS-QUEUES>`: Restrict the queue names a task may be
+  enqueued to. By default, only the ``"default"`` queue is allowed. Queue name
+  validation can be disabled by setting this to an empty list.
+
+Asynchronous support
+--------------------
+
+Django has developing support for asynchronous task backends.
+
+``django.tasks.backends.base.BaseTaskBackend`` has async variants of all base
+methods. By convention, the asynchronous versions of all methods are prefixed
+with ``a``. The arguments for both variants are the same.
+
+Retrieving backends
+-------------------
+
+Backends can be retrieved using the ``tasks`` connection handler::
+
+    from django.tasks import tasks
+
+    tasks["default"]  # The default backend
+    tasks["reserve"]  # Another backend
+
+The "default" backend is available as ``default_task_backend``::
+
+    from django.tasks import default_task_backend
+
+Backend features
+----------------
+
+Django's tasks framework is designed to be backend-agnostic. This means some
+backends may not implement the same features as others, such as executing ``async``
+functions.
+
+Depending on the backend configured, it may be necessary to gracefully degrade
+functionality, or use a system check to prevent the application from starting at all.
+
+To facilitate this, certain features can be checked on a backend:
+
+* :attr:`supports_defer <django.tasks.backends.base.BaseTaskBackend.supports_defer>`:
+  Can tasks be executed after a specific time using ``run_after``?
+
+* :attr:`supports_async_task <django.tasks.backends.base.BaseTaskBackend.supports_async_task>`:
+  Can ``async`` functions (coroutines) be used as task functions?
+
+* :attr:`supports_get_result <django.tasks.backends.base.BaseTaskBackend.supports_get_result>`:
+  Can a task's results be retrieved from another thread or process?
+
+Defining tasks
+==============
+
+Tasks are defined using the :meth:`django.tasks.task` decorator on a
+module-level function::
+
+    from django.tasks import task
+
+
+    @task()
+    def calculate_meaning_of_life(answer=42):
+        return answer
+
+Returned in a :class:`django.tasks.Task` instance.
+
+.. note::
+
+    All arguments and return values must be JSON serializable, and round-trip
+    as the same types.
+
+    This means complex types like model instances, as well as many built-in types
+    like ``datetime`` and ``tuple`` cannot be used in tasks without additional
+    conversion.
+
+The ``task`` decorator accepts a few keyword arguments to customize the task:
+
+* ``priority``: The priority of the task. Higher numbers will be executed sooner.
```

Comment:
I see a couple of options:

1. **Punt this to external tooling**: Use a management command that runs a scheduler. This is similar to how we would run Celery Beat in addition to a Celery worker, where the scheduler management command would, in addition to scheduling tasks, have the ability to reorder tasks in the queue. For instance, it would be fairly straightforward to use the [schedule](https://github.com/dbader/schedule) library with registered tasks and a management command that runs `schedule.run_pending()` in a loop (as an example to illustrate how it might work).

2. **Middleware pattern**: Implement something that would be similar in pattern to a "middleware for task requests" to allow for pluggable guard rails or other functionality. This would be more of a supplement to allow external tooling to handle more complex needs. For example, a common issue with schedulers is if the worker is offline but the scheduler keeps adding tasks to the queue indefinitely. There could be middleware-like functions that help act as circuit breakers, or often in simple projects we just want to ensure only one instance of each task is running at any time.

Scheduling does seem like an area that will quickly grow beyond the scope of what `django.tasks` should provide out of the box (though we can provide a reference implementation).

There is a lot of potential here if a small team can build out a set of background tasks in Django, and then plug an external scheduler or orchestrator into that. Imagine gaining the benefit of having your Django tasks orchestrated by Airflow, or similarly mature tooling that has already solved many challenges around scheduling and task orchestration involving complex task dependencies.

Reply:
Sorry, I am failing to see the link between your reply and my original message here. 

I believe here `priority` refers to a way that you can queue up messages into a same queue with different priorities, and that ends up with queues reordering messages to take the important ones "first".

Regarding scheduled tasks, I think there's a very large universe of possibilities like what you're mentioning. A problem where it's possible to have opinions, but definitely not a problem where there are no good solutions!

Reply:
You mentioned not including the priority feature if it were not already included, and the range of priorities possible, which suggests to me each project may have different ways of implementing the concept of prioritization.

Which raises the question, should this be something that’s natively a property of the task, versus something that is pluggable, i.e. handled by an external scheduled/orchestrator or implemented as a kind of middleware (not a Django middleware, but the same idea, specific to the task request/response cycle).

It doesn’t bother me personally if `priority` is directly on the task, but I foresee a need for many other customization needs (of which priority is one example), and middleware seems like a pattern that’s proven effective at customizing a request/response cycle, and well understood in the Django ecosystem.

Reply:
Have you seen any queues which don't have a concept of task priority? Sure, it might not be numeric, but casting them to smaller ranges should work.

Reply:
A simple database backend without a priority is straightforward. One with a priority number associated to tasks must now answer operational questions about how the priority is handled, especially if a single worker is listening to multiple queues and where the queue hygiene is not great (like some tasks take much longer than others).

I understand having the priority number be available because if you're using something like rabbit it's just there. I just think that a `DatabaseBackend` that ships with Django should be very simple if possible, and priorities feels like something that would require more finessing.

Overall thing here though: if we can have the `DatabaseBackend` in this PR then it would be easier to confirm/deny my suspicion on that point. If the code is already present then that's a fair argument against my critique!

Reply:
I will restate that I think including the DatabaseBackend in the first iteration is worth it, because currently this PR doesn't include any backends that run things in other processes. There are several things where discussions are happening abstractly but if we have the backend included they can be concrete.

An alternative idea would be to at least put up a WIP branch with the DB backend somewhere, to show how things will work in practice. I think it's important to show a successful implementation.

Or should I just be looking at django-tasks for the implementation there?

Reply:
Just look at `django-tasks`. Even then, it's _an_ implementation. 3rd-party implementations are still a requirement, so this common API needs to be generally useful. A single additional implementation doesn't necessarily validate or invalidate an assumption here.